File: ProblemList/00826.py
# BST
class Solution:
    def maxProfitAssignment(self, difficulty: list[int], profit: list[int], worker: list[int]) -> int:
        mix_array = []
        for i in range(len(difficulty)):
            mix_array.append((difficulty[i],profit[i]))
        mix_array.sort(key=lambda x: (x[0],x[1]))
        
        cur_max_profit = 0
        i = 0
        while i < len(mix_array):
            if mix_array[i][1] < cur_max_profit:
                mix_array.pop(i)
                i -= 1
            else:
                cur_max_profit = mix_array[i][1]
            i += 1
        
        mix_arr_len = len(mix_array)
        ans = 0
        for i in worker:
            start = 0; end = mix_arr_len-1
            while start <= end:
                mid = start + (end-start)//2
                if i >= mix_array[mid][0]:
                    start = mid+1
                else:
                    end = mid-1
            if end != -1:
                ans += mix_array[end][1]
        return ans
            

File: ProblemList/00826.cpp
#include <vector>
#include <map>
using namespace std;
// o(n)
class Solution {
public:
    int maxProfitAssignment(vector<int>& difficulty, vector<int>& profit, vector<int>& worker) {
        // o(n)
        int n = difficulty.size(), max_diff = 0;
        for(auto& d: difficulty){
            max_diff = max(max_diff,d);
        }

        vector<int> mp(max_diff+1,0);
        for(int i=0; i<n; i++){
            mp[difficulty[i]] = max(mp[difficulty[i]],profit[i]);
        }

        for(int i=1; i<=max_diff; i++){
            mp[i] = max(mp[i],mp[i-1]);
        }

        int ans = 0;
        for(auto&w :worker){
            ans += w > max_diff ? mp.back() : mp[w];
        }
        return ans;
    }
};

// o(nlogn) binary search
class Solution {
public:
    int maxProfitAssignment(vector<int>& difficulty, vector<int>& profit, vector<int>& worker) {
        map<int,int> mp;
        for(int i=0; i<difficulty.size(); i++){
            mp[difficulty[i]] = max(mp[difficulty[i]],profit[i]);
        }

        int cur_max = 0;
        for(auto& item: mp){
            item.second = max(cur_max,item.second);
            cur_max = max(cur_max,item.second);
        }
        sort(difficulty.begin(),difficulty.end());

        int ans = 0;
        for(auto& w: worker){
            int start = 0, end = difficulty.size()-1;
            while(start <= end){
                int mid = start + (end-start)/2;
                if(w >= difficulty[mid]) start = mid+1;
                else end = mid-1;
            }
            if(end >= 0) ans += mp[difficulty[end]];
        }
        return ans;
    }
};